File: data_store.py
#TODO return this in iter methods so they can be chained?
#TODO cache RowIter commands like only and exclude so that they aren't computed for every row
import pprint

# util methods

def _as_list(val):
    # convert single string to list
    return [val] if isinstance(val, basestring) else val


class RowIter(object):
    def __init__(self, table_obj):
        self._table_obj = table_obj
        self._record_index_selection = xrange(len(self._table_obj))
        self._sort_order = None

        self._reset()

    def _reset(self):
        self._iteration_count = 0

    #TODO queue these operations to ensure that a select_by/where always occurs
    # before an order_by

    def where(self, col_name, val):
        indexed_col_selection = self._table_obj._indexes.get(col_name)

        if indexed_col_selection is None:
            indexes = {}
            col_index = self._table_obj._col_name_index_map[col_name]
            for record_index in xrange(len(self._table_obj)):
                col_val = self._table_obj._data[record_index][col_index]
                indexes.setdefault(col_val, []).append(record_index)

            self._table_obj._indexes[col_name] = indexed_col_selection = indexes

        self._record_index_selection = indexed_col_selection.get(val, [])
        self._sort_order = None # any previous ordering is useless now

    def order_by(self, col_name):

        sort_order = []
        col_index = self._table_obj._col_name_index_map[col_name]
        for record_index in self._record_index_selection:
            sort_order.append((self._table_obj._data[record_index][col_index], record_index))
        sort_order.sort()
        self._sort_order = sort_order

    def __len__(self):
        return len(self._table_obj)

    def __iter__(self):
        return self

    def next(self):
        if self._iteration_count >= len(self._record_index_selection):
            raise StopIteration
        self._iteration_count += 1
        iteration = self._iteration_count - 1
        record_index = self._record_index_selection[iteration] if self._sort_order is None else self._sort_order[iteration][1]
        return ColIter(self._table_obj, record_index)

    # convenience methods

    def as_list(self, rows_as_lists=False):
        #TODO reset self._iteration_count?
        return [col_iter.as_list() if rows_as_lists else col_iter.as_dict() for col_iter in self]

    def as_dict(self, col_as_key, rows_as_lists=False):
        #TODO reset self._iteration_count?
        return {col_iter.get_by_name(col_as_key): (col_iter.as_list() if rows_as_lists else col_iter.as_dict()) for col_iter in self}

    # TODO port get_col from JS


class ColIter(object):
    def __init__(self, table_obj, record_index):
        self._table_obj = table_obj
        self._record_index = record_index
        self._record = table_obj._data[record_index]
        self._filtered_col_indexes = xrange(len(table_obj._col_defs))

        self._reset()

    def _reset(self):
        self._iteration_count = 0

    def only(self, col_names_list):
        col_names_list = _as_list(col_names_list)
        self._filtered_col_indexes = sorted(self._table_obj._col_name_index_map[col_name] for col_name in col_names_list)

    def exclude(self, col_names_list):
        #TODO slightly different approach in JS - iterated over mapped table._col_defs. re-port?
        col_names_list = _as_list(col_names_list)
        self._filtered_col_indexes = sorted(col_index for col_name, col_index in self._table_obj._col_name_index_map.items() if col_name not in col_names_list)

    def __len__(self):
        return len(self._filtered_col_indexes)

    def __iter__(self):
        return self

    def next(self):
        if self._iteration_count >= len(self._filtered_col_indexes):
            raise StopIteration
        self._iteration_count += 1

        # TODO re-work this to include helper method from JS (re-port from JS)

        col_index = self._filtered_col_indexes[self._iteration_count - 1]
        col_def = self._table_obj._col_defs[col_index]
        col_name = col_def['name']
        col_fk_table = col_def.get('fk_table')
        col_val = self._record[col_index]

        use_val = ColIter(col_fk_table, col_val) if ((col_fk_table is not None) and (col_val is not None)) else col_val

        return (col_name, use_val)


    def get_related(self, table_name, fk_col_name):
        related_table_obj = self._table_obj._related_fk[table_name][fk_col_name]
        row_iter = RowIter(related_table_obj)
        row_iter.where(fk_col_name, self._record_index)
        return row_iter

    # convenience methods

    def get_by_index(self, col_index):
        #TODO implement handling of FKs. re-port from JS
        return self._record[col_index]

    def get_by_name(self, col_name):
        #TODO implement handling of FKs. re-port from JS
        #TODO implement deep lookup of values in FKs
        col_index = self._table_obj._col_name_index_map[col_name]
        return self._record[col_index]

    def as_list(self):
        #TODO reset self._iteration_count?
        row_list = []
        for col_name, val in self:
            if type(val) is self.__class__:
                row_list.append(val.as_list())
            else:
                row_list.append(val)
        return row_list

    def as_dict(self):
        #TODO reset self._iteration_count?
        row_dict = {}
        for col_name, val in self:
            if type(val) is self.__class__:
                row_dict[col_name] = val.as_dict()
            else:
                row_dict[col_name] = val
        return row_dict


class Table(object):
    def __init__(self, name, load_data):
        self.name = name

        if type(load_data) is dict:
            self._col_defs_raw = load_data['columns']
            self._data = load_data.get('data', [])
        else:
            self._col_defs_raw = load_data
            self._data = []

        # {TABLE_NAME: {COL_NAME: table_obj} }
        self._related_fk = {}

        # {COL_NAME: {val:[pks..]}}
        self._indexes = {}

        # {record_hash: record_index}
        self._hash_cache = None

        self._hash_key_col_index = None

    def __len__(self):
        return len(self._data)

    def init(self, table_dict):
        col_name_index_map = {}
        normalized_col_defs = []

        for i, col_def in enumerate(self._col_defs_raw):
            col_def_dict = {}
            if type(col_def) is dict:
                col_name = col_def['name']
                col_def_dict['name'] = col_name
                fk_table_name = col_def.get('fk')
                if fk_table_name is not None:
                    fk_table = table_dict[fk_table_name]
                    col_def_dict['fk_table'] = fk_table
                    fk_table._related_fk.setdefault(self.name, {})[col_name] = self
                #TODO more work on this hash_key stuff. port to JS
                if col_def.get('hash_key', False):
                    self._hash_key_col_index = i
            else:
                col_def_dict['name'] = col_def

            normalized_col_defs.append(col_def_dict)
            col_name_index_map[col_def_dict['name']] = i

        self._col_name_index_map = col_name_index_map
        self._col_defs = normalized_col_defs
        del self._col_defs_raw

    @property
    def schema(self):
        #TODO add hash_key to schema and port to JS
        schema = []
        for col_def in self._col_defs:
            name = col_def['name']
            fk_table = col_def.get('fk_table')
            if fk_table is None:
                schema.append(name)
            else:
                schema.append({'name': name, 'fk': fk_table.name})
        return schema

    def serialize(self):
        return {'columns': self.schema, 'data': self._data}

    def _get_hash(self, raw_record):
        #TODO hash hints here
        #TODO what if hash_key is missing or None?

        if self._hash_key_col_index is not None:
            return raw_record[self._hash_key_col_index]
        else:
            return tuple(raw_record)

    def _build_hash_cache(self):
        self._hash_cache = {self._get_hash(raw_record): record_index for record_index, raw_record in enumerate(self._data)}

    def insert(self, data):
        #TODO invalidate caches
        raw_record = []
        for col_def in self._col_defs:
            col_name = col_def['name']
            val = data.get(col_name)
            fk_table = col_def.get('fk_table')
            if fk_table is None:
                raw_record.append(val)
            elif val is None:
                raw_record.append(None)
            else:
                fk_insert_index = fk_table.insert(val)
                raw_record.append(fk_insert_index)

        raw_record_hash = self._get_hash(raw_record)
        if self._hash_cache is None:
            self._build_hash_cache()
        existing_record_index = self._hash_cache.get(raw_record_hash)
        if existing_record_index is None:
            new_index = len(self)
            self._data.append(raw_record)
            self._hash_cache[raw_record_hash] = existing_record_index = new_index
        return existing_record_index


class DataStore(object):

    #TODO make sure/enforce all table names in a dataset are unique
    def __init__(self, load_data):
        self._table_dict = {}

        for table_name, table_data in load_data.items():
            table = Table(table_name, table_data)
            self._table_dict[table.name] = table

        for table in self._table_dict.values():
            table.init(self._table_dict)

    @property
    def schema(self):
        return {table_name: table_obj.schema for table_name, table_obj in self._table_dict.items()}

    def serialize(self):
        return {table_name: table_obj.serialize() for table_name, table_obj in self._table_dict.items()}

    def select(self, table_name):
        return RowIter(self._table_dict[table_name])

    def get_record(self, table_name, record_index):
        return ColIter(self._table_dict[table_name], record_index)

    #TODO port lookup from js

    def insert(self, table_name, data):
        return self._table_dict[table_name].insert(data)

    def insert_many(self, table_name, data_list):
        table_obj = self._table_dict[table_name]
        for data_dict in data_list:
            table_obj.insert(data_dict)
        #TODO return a list of inserted indexes here? Not sure why we would ever need this...

File: static/js/data_store.js
'use strict'

//TODO experiment with using prototypes for common methods instead of member methods for performance in the future

function DataStore(load_data){

    // util functions

    function each(obj, callback){
        for(const key in obj){
            if(obj.hasOwnProperty(key)){
                callback(key, obj[key])
            }
        }
    }

    function map(obj, callback){
        let return_dict = {}
        for(const key in obj){
            if(obj.hasOwnProperty(key)){
                let cb_val = callback(key, obj[key])
                return_dict[cb_val[0]] = cb_val[1]
            }
        }
        return return_dict
    }

    function setdefault(obj, key, default_val){
        let val = obj[key]
        if(val === undefined){
            obj[key] = val = default_val
        }
        return val
    }

    function as_list(val){
        // convert single string to list
        return Array.isArray(val) ? val : [val]
    }

    // helpers for RowIter and ColIter prototypes

    function _only(table_obj, col_names_list){
        col_names_list = as_list(col_names_list)
        let filtered_col_indexes = col_names_list.map(function(col_name){
            return table_obj._col_name_index_map[col_name]
        })
        filtered_col_indexes.sort()
        return filtered_col_indexes
    }

    function _exclude(table_obj, col_names_list){
        col_names_list = as_list(col_names_list)
        let filtered_col_indexes = []
        for(let col_index = 0, count = table_obj._col_defs.length; col_index < count; col_index++){
            let col_name = table_obj._col_defs[col_index]['name']
            if(col_names_list.indexOf(col_name) == -1){
                filtered_col_indexes.push(col_index)
            }
        }
        // sorting not necessary here because we iterated over table_obj._col_defs
        return filtered_col_indexes
    }

    function RowIter(table_obj){
        this._table_obj = table_obj
        this._record_index_selection = null
        this._sort_order = null
        this._filtered_col_indexes = null
        this._rows_as_lists = false
        this._one_col_name_array = null
    }

    // helper for RowIter prototypes
    //TODO port this to Python

    RowIter.prototype._get_col_iter_data = function(col_iter){
        if(this._one_col_name_array !== null){
            return col_iter.one(this._one_col_name_array)
        }

        return this._rows_as_lists ? col_iter.values() : col_iter.dict()
    }


    //methods 'pass-thru' to ColIter
    //TODO port these to Python

    RowIter.prototype.only = function(col_names_list){
        this._filtered_col_indexes = _only(this._table_obj, col_names_list)
        return this
    }

    RowIter.prototype.exclude = function(col_names_list){
        this._filtered_col_indexes = _exclude(this._table_obj, col_names_list)
        return this
    }

    RowIter.prototype.values = function(){
        this._rows_as_lists = true
        return this
    }

    RowIter.prototype.one = function(col_name_array){
        this._one_col_name_array = col_name_array
        return this
    }

    //TODO queue these operations to ensure that a select_by/where always occurs
    //before an order_by

    RowIter.prototype.where = function(col_name, val){
        let indexed_col_selection = this._table_obj._indexes[col_name]

        if(indexed_col_selection === undefined){
            let indexes = {}
            let col_index = this._table_obj._col_name_index_map[col_name]
            for(let record_index = 0, count = this._table_obj._data.length; record_index < count; record_index++){
                let col_val = this._table_obj._data[record_index][col_index]
                setdefault(indexes, col_val, []).push(record_index)
            }
            this._table_obj._indexes[col_name] = indexed_col_selection = indexes
        }

        this._record_index_selection = indexed_col_selection[val] || []
        this._sort_order = null // any previous ordering is useless now
        return this
    }

    RowIter.prototype.order_by = function(col_name, sort_numeric){
        this._sort_order = []
        let col_index = this._table_obj._col_name_index_map[col_name]
        let is_subset = this._record_index_selection !== null
        let use_count = is_subset ? this._record_index_selection.length : this._table_obj._data.length
        for(let iter_index = 0; iter_index < use_count; iter_index++){
            let record_index = is_subset ? this._record_index_selection[iter_index] : iter_index
            this._sort_order.push([this._table_obj._data[record_index][col_index], record_index])
        }

        if(sort_numeric){
            this._sort_order.sort(function(a, b){return a[0] - b[0]})
        }
        else{
            this._sort_order.sort()
        }

        return this
    }

    RowIter.prototype[Symbol.iterator] = function(){
        //TODO implement as generator to simplify this a bit?
        let table_obj = this._table_obj
        let iteration_count = 0
        let is_sorted = (this._sort_order !== null)
        let use_selection = is_sorted ? this._sort_order : this._record_index_selection
        let count = (use_selection === null) ? table_obj._data.length : use_selection.length
        let filtered_col_indexes = this._filtered_col_indexes

        return {
            'next': function(){
                if(iteration_count >= count){
                    return {'done': true}
                }
                let i = iteration_count++
                let record_index = (use_selection === null) ? i : use_selection[i]
                if(is_sorted){
                    record_index = record_index[1]
                }
                return {'value': new ColIter(table_obj, record_index, filtered_col_indexes), 'done': false}
            }
        }
    }

    RowIter.prototype.as_list = function(){
        let row_iter_list = []
        for(const col_iter of this){
            row_iter_list.push(this._get_col_iter_data(col_iter))
        }
        return row_iter_list
    }

    RowIter.prototype.as_dict = function(col_as_key){
        let row_iter_dict = {}
        for(const col_iter of this){
            row_iter_dict[col_iter.one(col_as_key)] = this._get_col_iter_data(col_iter)
        }
        return row_iter_dict
    }


    function ColIter(table_obj, record_index, filtered_col_indexes){
        this._table_obj = table_obj
        this._record_index = record_index
        this._record = table_obj._data[record_index]
        this._filtered_col_indexes = filtered_col_indexes || null
    }

    // helper for ColIter prototypes

    function _get_val_or_col_iter(table_obj, record, col_index){
        let val = record[col_index]
        let col_def = table_obj._col_defs[col_index]
        let col_fk_table = col_def['fk_table']
        if((col_fk_table !== undefined) && (val !== null)){
            val = new ColIter(col_fk_table, val)
        }
        return [col_def['name'], val]
    }

    ColIter.prototype[Symbol.iterator] = function(){
        //TODO implement as generator to simplify this a bit?
        let table_obj = this._table_obj
        let record = this._record
        let iteration_count = 0
        let filtered_col_indexes = this._filtered_col_indexes
        let is_filtered = (filtered_col_indexes !== null)
        let count = is_filtered ? filtered_col_indexes.length : table_obj._col_defs.length

        return {
            'next': function(){
                if(iteration_count >= count){
                    return {'done': true}
                }
                let i = iteration_count++
                let col_index = is_filtered ? filtered_col_indexes[i] : i
                let val = _get_val_or_col_iter(table_obj, record, col_index)
                return {'value': val, 'done': false}
            }
        }
    }

    ColIter.prototype.get_related = function(table_name, fk_col_name){
        let related_table_obj = this._table_obj._related_fk[table_name][fk_col_name]
        let row_iter = new RowIter(related_table_obj)
        row_iter.where(fk_col_name, this._record_index)
        return row_iter
    }

    ColIter.prototype.only = function(col_names_list){
        this._filtered_col_indexes = _only(this._table_obj, col_names_list)
        return this
    }

    ColIter.prototype.exclude = function(col_names_list){
        this._filtered_col_indexes = _exclude(this._table_obj, col_names_list)
        return this
    }


    ColIter.prototype.one = function(col_name_array){
        // col_name_array allows for deep lookups of nested FK records
        //TODO would this be better with dot notation? It would mean that we would have to enforce that
        // dots can't be in string column names
        col_name_array = as_list(col_name_array)

        let col_name_index = 0

        let col_index = this._table_obj._col_name_index_map[col_name_array[col_name_index]]
        let val = _get_val_or_col_iter(this._table_obj, this._record, col_index)[1]

        while(++col_name_index < col_name_array.length){
            val = val.one(col_name_array[col_name_index])
        }

        return val
    }

    ColIter.prototype.values = function(){
        let row_list = []
        for(const pair of this){
            let val = pair[1]
            row_list.push((val instanceof ColIter) ? val.as_list() : val)
        }
        return row_list
    }

    ColIter.prototype.dict = function(){
        let row_dict = {}
        for(const pair of this){
            let col_name = pair[0]
            let val = pair[1]
            row_dict[col_name] = (val instanceof ColIter) ? val.as_dict() : val
        }
        return row_dict
    }




    function Table(name, load_data){

        this._col_defs = []
        this._col_name_index_map = {}
        this._data = []

        // {TABLE_NAME: {COL_NAME: table_obj} }
        this._related_fk = {}

        // {COL_NAME: {val:[pks..]}}
        this._indexes = {}

        // {record_hash: record_index}
        let _hash_cache = null

        this.schema = []

        this.init = function(){
            let col_defs_raw
            if(Array.isArray(load_data)){
                col_defs_raw = load_data
            }
            else{
                col_defs_raw = load_data['columns']
                this._data = load_data['data'] || []
            }

            for(let i=0; i < col_defs_raw.length; i++){
                let col_def = col_defs_raw[i]
                let col_def_dict = {}
                let col_name
                let schema
                if(typeof col_def === 'object'){ //TODO is a more robust check needed here?
                    col_name = col_def['name']
                    col_def_dict['name'] = col_name
                    let fk_table_name = col_def['fk']
                    if(fk_table_name !== undefined){
                        let fk_table = _table_dict[fk_table_name]
                        col_def_dict['fk_table'] = fk_table
                        setdefault(fk_table._related_fk, name, {})[col_name] = this
                        schema = {'name': col_name, 'fk': fk_table_name}
                    }
                }
                else{
                    col_def_dict['name'] = col_name = col_def
                }

                this._col_defs.push(col_def_dict)
                this._col_name_index_map[col_name] = i

                this.schema.push(schema || col_name)
            }
        }

        this.serialize = function(){
            return {'columns': this.schema, 'data': this._data}
        }


    }

    let _table_dict = {}

    this.schema = {}

    const self = this

    each(load_data, function(table_name, table_data){
        let table = new Table(table_name, table_data)
        _table_dict[table_name] = table
    })

    each(_table_dict, function(table_name, table){
        table.init()
        self.schema[table_name] = table.schema
    })

    this.serialize = function(){
        return map(_table_dict, function(table_name, table_obj){
            return [table_name, table_obj.serialize()]
        })
    }

    this.select = function(table_name){
        return new RowIter(_table_dict[table_name])
    }

    this.get_record = function(table_name, record_index){
        return new ColIter(_table_dict[table_name], record_index)
    }

    this.lookup_record = function(table_name, col_name, val){
        let row_iter = new RowIter(_table_dict[table_name])
        row_iter.where(col_name, val)
        for(const col_iter of row_iter){
            return col_iter
        }
        return null
    }

    this.insert = function(table_name, data){
        return _table_dict[table_name].insert(data)
    }

    this.insert_many = function(table_name, data_list){
        let table_obj = _table_dict[table_name]
        data_list.forEach(function(data_dict){
            table_obj.insert(data_dict)
            //TODO return a list of inserted indexes here? Not sure why we would ever need this...
        })
    }
}
